theforeman.org/pipelines/test/testKatello.groovy

```diff
@@ -1,175 +1,163 @@
+                            ghprb_git_checkout()
 
-                configureRVM(ruby)
+                            dir('foreman') {
+                               git url: "https://github.com/theforeman/foreman", branch: foreman_branch, poll: false
+                            }
+                        }
+                    }
+                    stage("Setup RVM") {
+                        steps {
 
-            }
-        }
-        stage('Configure Environment') {
-            steps {
+                            configureRVM(ruby)
 
-                dir('foreman') {
-                    addGem()
-                    databaseFile(gemset())
-                }
+                        }
+                    }
+                    stage('Configure Environment') {
+                        steps {
 
-            }
-        }
-        stage('Configure Database') {
-            steps {
+                            dir('foreman') {
+                                addGem()
+                                databaseFile(gemset())
+                            }
 
-                dir('foreman') {
-                    configureDatabase(ruby)
-                }
+                        }
+                    }
+                    stage('Configure Database') {
+                        steps {
+
+                            dir('foreman') {
+                                configureDatabase(ruby)
+                            }
 
-            }
-        }
-        stage('Install Foreman npm packages') {
-            steps {
-                dir('foreman') {
-                    withRVM(["bundle exec npm install"], ruby)
-                }
-            }
-        }
-        stage('Run Tests') {
-            parallel {
-                stage('tests') {
-                    steps {
-                        dir('foreman') {
-                            withRVM(['bundle exec rake jenkins:katello TESTOPTS="-v" --trace'], ruby)
                         }
                     }
-                }
-                stage('rubocop') {
-                    steps {
-                        dir('foreman') {
-                            withRVM(['bundle exec rake katello:rubocop TESTOPTS="-v" --trace'], ruby)
+                    stage('Install Foreman npm packages') {
+                        steps {
+                            dir('foreman') {
+                                withRVM(["bundle exec npm install"], ruby)
+                            }
                         }
                     }
-                }
-                stage('react-ui') {
-                    when {
-                        expression { fileExists('package.json') }
+                    stage('rubocop') {
+                        steps {
+                            dir('foreman') {
+                                withRVM(['bundle exec rake katello:rubocop TESTOPTS="-v" --trace'], ruby)
+                            }
+                        }
                     }
```

Comment:
Maybe not something you want to change in this PR, but I'd recommend running Rubocop only once (on the latest version). Also consider publishing the results like we do in smart-proxy:
https://github.com/theforeman/jenkins-jobs/blob/ed25297a67807f6d5684c57ab42cd6fde75a8627/theforeman.org/pipelines/test/smart-proxy.groovy#L35-L43

Note that it takes the approach of running rubocop in the tool itself rather than via the common rake job via Foreman which may or may not be what you want.

I'd also run rubocop before running `npm install` so you fail fast. In fact, I would suggest to run rubocop at the very beginning so you execute as little as possible if rubocop fails.

Reply:
That sounds like a good follow up given I need to make sure it all works and I am contemplating some further improvements to break out the design.